File: files/2025-06-01/CSEC 01062025.R
library(tidyverse)
library(readxl)
library(lubridate)

path = "files/2025-06-01/Challenge 30.xlsx"
start_date = as.Date("2025-05-01")
work_days = 8
travel_days = 8
test = read_excel(path, range = "F3:G23") %>%
  mutate(Schedule = as.Date(Schedule))

df = data.frame(
  dates = seq.Date(
    from = start_date,
    by = "day",
    length.out = (work_days + travel_days) * 2
  )
)
df1 = df %>%
  mutate(weekend = ifelse(wday(dates) %in% c(7, 1), TRUE, FALSE)) %>%
  mutate(weekend = ifelse(weekend, "Rest", "Work")) %>%
  mutate(work_count = row_number(), .by = weekend) %>%
  mutate(
    work_index = ifelse(
      weekend == "Work" & work_count == work_days + 1,
      row_number(),
      NA
    )
  ) %>%
  fill(work_index, .direction = "downup") %>%
  mutate(weekend = ifelse(row_number() >= work_index, "Travel", weekend)) %>%
  mutate(travel_count = row_number(), .by = weekend) %>%
  filter(travel_count <= travel_days) %>%
  select(Schedule = dates, Activity = weekend)

all.equal(df1, test, check.attributes = FALSE)
# [1] TRUE

File: files/2025-06-01/CSEC 01062025.py
import pandas as pd
from datetime import datetime, timedelta

path = "files/2025-06-01/Challenge 30.xlsx"
start_date = pd.to_datetime("2025-05-01")
work_days = 8
travel_days = 8

test = pd.read_excel(path, usecols="F:G", skiprows=2, nrows=21)
test['Schedule'] = pd.to_datetime(test['Schedule'])

dates = pd.date_range(start_date, periods=(work_days + travel_days) * 2)
df = pd.DataFrame({'dates': dates})
df['weekend'] = df['dates'].dt.weekday.isin([5, 6])
df['weekend'] = df['weekend'].map({True: 'Rest', False: 'Work'})
df['work_count'] = df.groupby('weekend').cumcount() + 1

df['work_index'] = None
mask = (df['weekend'] == 'Work') & (df['work_count'] == work_days + 1)
if mask.any():
    idx = df.index[mask][0]
    df.loc[idx:, 'work_index'] = idx
df['work_index'] = df['work_index'].ffill().bfill()

df.loc[df.index >= df['work_index'], 'weekend'] = 'Travel'
df['travel_count'] = df.groupby('weekend').cumcount() + 1
df1 = df[df['travel_count'] <= travel_days][['dates', 'weekend']]
df1.columns = ['Schedule', 'Activity']

print(df1.equals(test))